File: 05-adversarial/src/ledger.ts
// 05-adversarial/src/ledger.ts
// A minimal points ledger. Accounts hold points, and transfer moves points between accounts.
// No framework, no build step required.

export class Ledger {
  private balances = new Map<string, number>();

  constructor(initial: Record<string, number> = {}) {
    for (const [account, points] of Object.entries(initial)) {
      this.balances.set(account, points);
    }
  }

  balanceOf(account: string): number {
    return this.balances.get(account) ?? 0;
  }

  transfer(from: string, to: string, amount: number): void {
    const available = this.balanceOf(from);
    if (available < amount) {
      throw new Error("insufficient balance");
    }
    this.balances.set(from, available - amount);
    this.balances.set(to, this.balanceOf(to) + amount);
  }
}

File: 05-adversarial/src/ledger.py
# 05-adversarial/src/ledger.py
# A minimal points ledger. Accounts hold points, and transfer moves points between accounts.
# No framework required.


class Ledger:
    def __init__(self, initial=None):
        self._balances = dict(initial or {})

    def balance_of(self, account):
        return self._balances.get(account, 0)

    def transfer(self, sender, recipient, amount):
        available = self.balance_of(sender)
        if available < amount:
            raise ValueError("insufficient balance")
        self._balances[sender] = available - amount
        self._balances[recipient] = self.balance_of(recipient) + amount
